### PR TITLE
Packed QKV and Rotary Embedding Support for sm<80 GQA

### PR DESCRIPTION
**Pull request by aciddelgado**
### Description
Add support for packed qkv input and rotary embedding with sm<80 using memory efficient attention kernel.

### Motivation and Context
Allows lower-end gpus to run gqa with packed qkv input and rotary embedding.

### CODE REVIEW

File: onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc [hunk 1]
```diff
@@ -251,7 +262,13 @@ Status GroupQueryAttention<T>::ComputeInternal(OpKernelContext* context) const {
   if (fmha_buffer != nullptr) {
     data.fmha_buffer = reinterpret_cast<CudaT*>(fmha_buffer.get());
   }
-  // Rotary
+  if (packed_qkv_buffer != nullptr) {
+    data.packed_qkv = reinterpret_cast<CudaT*>(packed_qkv_buffer.get());
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc:266]: May use name like unpacked_qkv, unpacked_qkv_buffer

File: onnxruntime/contrib_ops/cuda/bert/group_query_attention_impl.cu [hunk 2]
```diff
@@ -563,9 +642,49 @@ Status EfficientAttention(
...
+    key = reinterpret_cast<const void*>(data.key);
+    value = reinterpret_cast<const void*>(data.value);
+  } else {
+    size_t q_size = static_cast<size_t>(batch_size * sequence_length * num_heads * head_size);
+    size_t k_size = static_cast<size_t>(batch_size * sequence_length * kv_num_heads * head_size);
+    auto q = reinterpret_cast<T*>(data.packed_qkv);
+    auto k = reinterpret_cast<T*>(data.packed_qkv + q_size);
+    auto v = reinterpret_cast<T*>(data.packed_qkv + q_size + k_size);
+    ORT_RETURN_IF_ERROR(LaunchUnpackQKV(reinterpret_cast<const T*>(data.query), q, k, v, num_heads, kv_num_heads, head_size,
+                        sequence_length, batch_size, stream, max_threads_per_block));
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/bert/group_query_attention_impl.cu:660]: Not aligned. Try clang format.

File: onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc [hunk 3]
```diff
@@ -172,18 +170,31 @@ Status GroupQueryAttention<T>::ComputeInternal(OpKernelContext* context) const {
   if (use_memory_efficient_attention && needs_buff) {
     kv_buffer_bytes = (sizeof(T) * parameters.batch_size * parameters.num_heads * parameters.seqlen_present_kv_cache * parameters.head_size);
   }
+  size_t rotary_buffer_bytes = 0;
+  if (use_memory_efficient_attention && do_rotary_) {
```
Comment (kunal-vaishnavi) [inline onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc:174]: If memory efficient attention and flash attention are disabled/unused and GroupQueryAttention + RotaryEmbedding fusion is done, will GQA with `do_rotary` enabled be able to run? If not, I think we should document this limitation.

File: onnxruntime/contrib_ops/cuda/bert/group_query_attention_impl.cu [hunk 4]
```diff
@@ -563,9 +642,49 @@ Status EfficientAttention(
...
+    value = reinterpret_cast<const void*>(v);
+  }
+
+  if (parameters.do_rotary) {
+    size_t q_size = static_cast<size_t>(batch_size * sequence_length * num_heads * head_size);
+    size_t k_size = static_cast<size_t>(batch_size * sequence_length * kv_num_heads * head_size);
+    auto q_buffer = reinterpret_cast<T*>(data.rotary_buffer);
+    auto k_buffer = q_buffer + q_size;
+    auto position_ids_buff = reinterpret_cast<int64_t*>(k_buffer + k_size);
+    // std::cout << parameters.seqlen_present_kv_cache << std::endl;
```
Comment (kunal-vaishnavi) [inline onnxruntime/contrib_ops/cuda/bert/group_query_attention_impl.cu:672]:
```suggestion
```

File: onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc (hunk 3 above)
Comment (aciddelgado) [reply onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc:174]: there is no unfused gqa implementation at all actually... should i document this detail in bert_defs?

Comment (kunal-vaishnavi) [reply onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc:174]: Yes, let's document it.

**Review by tianleiwu — APPROVED** [review]

**Review by tianleiwu — APPROVED** [review]

**Pull request merged**